pulpcore/pulpcore/app/models/repository.py

```diff
@@ -221,25 +207,147 @@ class RepositoryContent(Model):
 
         content (models.ForeignKey): The associated content.
         repository (models.ForeignKey): The associated repository.
+        vadded (models.ForeignKey): The RepositoryVersion which added the referenced Content.
+        vremoved (models.ForeignKey): The RepositoryVersion which removed the referenced Content.
     """
-    content = models.ForeignKey('Content', on_delete=models.CASCADE)
+    content = models.ForeignKey('Content', on_delete=models.CASCADE,
+                                related_name='version_memberships')
     repository = models.ForeignKey(Repository, on_delete=models.CASCADE)
+    vadded = models.ForeignKey('RepositoryVersion', related_name='added_memberships')
+    vremoved = models.ForeignKey('RepositoryVersion', null=True, related_name='removed_memberships')
+
+    class Meta:
+        unique_together = (('repository', 'content', 'vadded'),
+                           ('repository', 'content', 'vremoved'))
+
+
+class RepositoryVersion(Model):
+    """
+    A version of a repository's content set.
+    Fields:
+        number (models.PositiveIntegerField): A positive integer that uniquely identifies a version
+            of a specific repository. Each new version for a repo should have this field set to
+            1 + the most recent version.
+        created (models.DateTimeField): When the version was created.
+        action  (models.TextField): The action that produced the version.
+    Relations:
+        repository (models.ForeignKey): The associated repository.
+    """
+    SYNC = 'sync'
+    COPY = 'copy'
+    UPLOAD = 'upload'
+    ADD = 'add'
+    REMOVE = 'remove'
+    ACTIONS = (
+        (SYNC, 'Sync'),
+        (COPY, 'Copy'),
+        (UPLOAD, 'Upload'),
+        (ADD, 'Add'),
+        (REMOVE, 'Remove'),
+    )
+
+    repository = models.ForeignKey(Repository)
```

Comment:
Also the MVP calls out removing the association of an importer with any specific repo, and replace it with an attribute on a repo called `default_importer` that is a foreign key to a single importer. A post can specify an 'importer' setting which tells the sync code to call that importer to finish the repo version. If 'importer' is not specified, but the associated importer has `default_importer` set, then use that as the importer.

Reply:
Lastly there is an idea to use a `base_version`. This `base_version` would be an option the user could specify. If unspecified it defaults to the latest version of the associated repo. If specified as None then the new version starts with no content. If specified as a href to a version, the repo version begins with the content from that version before applying add_many or remove_many or calling into an importer.

Reply:
> Also the MVP calls out removing the association of an importer with any specific repo, and replace it with an attribute on a repo called default_importer that is a foreign key to a single importer. A post can specify an 'importer' setting which tells the sync code to call that importer to finish the repo version. If 'importer' is not specified, but the associated importer has default_importer set, then use that as the importer.

Can you elaborate on how that change is relevant to this code? I'm just not seeing the connection.

Reply:
> Lastly there is an idea to use a base_version. This base_version would be an option the user could specify. If unspecified it defaults to the latest version of the associated repo. If specified as None then the new version starts with no content. If specified as a href to a version, the repo version begins with the content from that version before applying add_many or remove_many or calling into an importer.

Interesting idea. What are the use cases motivating it?

I assume that behavior would mostly happen in task code, and not here on the model, right?

And would this be a 3.1+ idea?